src/gordon_gcp/plugins/publisher.py

```diff
@@ -43,14 +57,200 @@ class GDNSPublisher:
+        _api_version = config.get('version', 'v1')
+        self.base_url = BASE_CHANGES_ENDPOINT.format(
+            host=HOST, version=_api_version, project=self.project,
+            managedZone=self.managed_zone)
+        self._logger = logging.getLogger('')
+
+    # TODO: This will be eventually moved to GEventMessage
+    async def update_phase(self, event_msg, phase=None):
+        old_phase = event_msg.phase
+        event_msg.phase = phase or self.phase
+        msg = f'Updated phase from "{old_phase}" to "{event_msg.phase}".'
+        event_msg.append_to_history(msg, self.phase)
+
+    def _format_changes(self, action, resource_record):
+        """Return dict containing the changes to be made.
+
+        Args:
+            action (str): the action of the changes to publish,
+                can be one of ['additions', 'deletions'].
+            resource_record (dict): dict contains record data.
+
+        Returns:
+            The changes (dict) to post to google API.
+        """
+        resource_record['kind'] = 'dns#resourceRecordSet'
+        resource_record['ttl'] = resource_record.get('ttl', self.default_ttl)
+
+        return {
+            'kind': 'dns#change',
+            action: [resource_record]
+        }
+
+    async def _watch_status(self, changes_id):
+        """Check google API if the changes are done.
+
+        Args:
+            changes_id (int): check the status for the changes_id.
+
+        Returns:
+            True if the changes are done.
+
+        Raises:
+            Raise error if timeout passed and
+                we stiLl didn't get status "Done" from the API.
+        """
+        url = f'{self.base_url}/{changes_id}'
+
+        timeout = datetime.timedelta(seconds=self.timeout)
+        start = datetime.datetime.now()
+        end = start + timeout
+
+        while datetime.datetime.now() < end:
+            resp = await self.http_client.get_json(url)
+
+            if resp['status'] == 'done':
+                return True
+
+            await asyncio.sleep(1)
+
+        msg = 'Timed out waiting for DNS changes to be done'
+        raise exceptions.GCPRetryMessageError(msg)
+
+    async def _publish_changes(self, changes, rec=False):
+        """Post changes to the google API and watch request status.
+
+        Args:
+            changes (dict): the changes to make.
+            rec (bool): flag to call _handle_update only once.
+
+        Returns:
+            Boolean if the changes are done.
+        """
+        try:
+            resp = await self.http_client.request('post', self.base_url,
+                                                  json=changes)
+        except Exception as e:
+            error_msg = e.args[0]
+            # extract status code from error message
+            match = re.search('\d+', error_msg)
+            if match:
+                status_code = int(match.group(0))
+
+                if status_code == 409 and not rec:
```

Comment:
would `_handle_update` ever be called? I can see a 409 being raised, and `rec` set to `False` by default so this would never actually handle the update, unless I'm not following this correctly?